codegenerator/cli/templates/dynamic/codegen/src/db/Entities.res.hbs

```diff
@@ -23,6 +23,240 @@ let isIndex = true
+
+    let previousHistoryFields =
+      previousChangeFieldNames->Belt.Array.map(fieldName =>
+        mkField(fieldName, Integer, ~isNullable)
+      )
+
+    let id = "id"
+
+    let dataFields = table.fields->Belt.Array.keepMap(field =>
+      switch field {
+      | Field(field) =>
+        switch field.fieldName {
+        //id is not nullable and should be part of the pk
+        | "id" => {...field, fieldName: id, isPrimaryKey: true}->Field->Some
+        //db_write_timestamp can be removed for this. TODO: remove this when we depracate
+        //automatic db_write_timestamp creation
+        | "db_write_timestamp" => None
+        | _ =>
+          {
+            ...field,
+            isNullable: true, //All entity fields are nullable in the case
+            isIndex: false, //No need to index any additional entity data fields in entity history
+          }
+          ->Field
+          ->Some
+        }
+
+      | DerivedFrom(_) => None
+      }
+    )
+
+    let dataFieldNames = dataFields->Belt.Array.map(field => field->getFieldName)
+
+    let originTableName = table.tableName
+    let historyTableName = originTableName ++ "_history"
+    //ignore composite indices
+    let table = mkTable(
+      historyTableName,
+      ~fields=Belt.Array.concatMany([currentHistoryFields, previousHistoryFields, dataFields]),
+    )
+
+    let createInsertFnQuery = {
+      let insertFnName = `"insert_${table.tableName}"`
+      let historRowArg = "history_row"
+      let historyTablePath = `"public"."${historyTableName}"`
+      let originTablePath = `"public"."${originTableName}"`
+
+      let previousHistoryFieldsAreNullStr =
+        previousChangeFieldNames
+        ->Belt.Array.map(fieldName => `${historRowArg}.${fieldName} IS NULL`)
+        ->Js.Array2.joinWith(" OR ")
+
+      let currentChangeFieldNamesCommaSeparated = currentChangeFieldNames->Js.Array2.joinWith(", ")
+
+      let dataFieldNamesDoubleQuoted = dataFieldNames->Belt.Array.map(fieldName => `"${fieldName}"`)
+      let dataFieldNamesCommaSeparated = dataFieldNamesDoubleQuoted->Js.Array2.joinWith(", ")
+
+      let allFieldNames = Belt.Array.concatMany([
+        currentChangeFieldNames,
+        previousChangeFieldNames,
+        dataFieldNamesDoubleQuoted,
+      ])
+
+      `CREATE OR REPLACE FUNCTION ${insertFnName}(${historRowArg} ${historyTablePath})
+      RETURNS void AS $$
+      DECLARE
+        v_previous_record RECORD;
+        v_origin_record RECORD;
+      BEGIN
+        -- Check if previous values are not provided
+        IF ${previousHistoryFieldsAreNullStr} THEN
+          -- Find the most recent record for the same id
+          SELECT ${currentChangeFieldNamesCommaSeparated} INTO v_previous_record
+          FROM ${historyTablePath}
+          WHERE ${id} = ${historRowArg}.${id}
+          ORDER BY ${currentChangeFieldNames
+        ->Belt.Array.map(fieldName => fieldName ++ " DESC")
+        ->Js.Array2.joinWith(", ")}
```

Comment:
I think it should take the previous one even without the ORDER BY

Reply:
Postgres doesn't guarentee ordering say by insertion. So what we should probably do is have a serial field that increments and we can order by that instead. What do you think? 

Reckon we will need that for the suggested unordered multichain implementation as well.

Reply:
> Postgres doesn't guarentee ordering say by insertion.

Didn't know that.

> So what we should probably do is have a serial field that increments and we can order by that instead. What do you think?

This sounds good, but the current approach also works

Reply:
It works for ordered multichain 😬